cmd/cloud-endpoints-controller/main.go

```diff
@@ -126,86 +130,54 @@ func sync(parent *CloudEndpoint, children *CloudEndpointControllerRequestChildre
-						svc, err := config.clientset.CoreV1().Services(parent.Spec.TargetIngress.Namespace).Get(svcName, metav1.GetOptions{})
-						if err != nil {
-							return status, &desiredChildren, fmt.Errorf("Failed to populate JWT audience from kubernetes service, not found: '%s', %v", svcName, err)
-						}
-						if svc.Spec.Type == corev1.ServiceTypeNodePort && len(svc.Spec.Ports) > 0 {
-							nodePort := strconv.Itoa(int(svc.Spec.Ports[0].NodePort))
-							found := false
-							for _, be := range ingBackends {
-								if strings.Contains(be, fmt.Sprintf("k8s-be-%s", nodePort)) {
-									bePatterns[i] = be
-									backend, err := config.clientCompute.BackendServices.Get(config.Project, be).Do()
-									if err == nil {
-										found = true
-										jwtAud := makeJWTAudience(config.ProjectNum, strconv.Itoa(int(backend.Id)))
-										log.Printf("[INFO][%s] Created jwtAud: %s", parent.Name, jwtAud)
-										status.JWTAudiences = append(status.JWTAudiences, jwtAud)
-									}
-								}
-							}
-							if found == false {
-								return status, &desiredChildren, fmt.Errorf("Backend not found or is not ready for service: %s, NodePort: %s", svcName, nodePort)
-							}
-						} else {
-							return status, &desiredChildren, fmt.Errorf("Service %s not type NodePort", svcName)
-						}
-					}
-				}
+		if name, key := parent.Spec.OpenAPISpecConfigMap.Name, parent.Spec.OpenAPISpecConfigMap.Key; name != "" && key != "" {
+			openAPISpecTemplate, err = getConfigMapSpecData(parent.ObjectMeta.Namespace, name, key)
+			if err != nil { //The user tried to supply a configMap spec, but it could not be loaded yet
+				log.Printf("[INFO][%s] Unable to load ConnfigMap Spec: %v", parent.Name, err)
+				return status, &desiredChildren, nil
 			} else {
-				target = parent.Spec.Target
-			}
-			openAPISpec, err = makeDefaultOpenAPISpec(status.Endpoint, target, status.JWTAudiences)
-			if err != nil {
-				return status, &desiredChildren, err
+				status.ConfigMapHash = toSha1(openAPISpecTemplate)
 			}
-			log.Printf("[INFO][%s] Target: %s", parent.Name, target)
-			status.IngressIP = target
-		} else {
-			// Use provided spec.
-			specYaml, err := yaml.Marshal(&parent.Spec.OpenAPISpec)
+		} else if parent.Spec.OpenAPISpec != nil {
+			openAPISpecTemplateBytes, err := yaml.Marshal(parent.Spec.OpenAPISpec)
 			if err != nil {
-				return status, &desiredChildren, err
+				log.Printf("[ERROR][%s] Supplied spec: %v", parent.Name, err)
+				return status, &desiredChildren, errors.Wrap(err, "Could not load user supplied API from CloudEndpoint Spec")
 			}
+			openAPISpecTemplate = string(openAPISpecTemplateBytes)
+		}
 
-			var spec map[string]interface{}
-			if err = yaml.Unmarshal([]byte(specYaml), &spec); err != nil {
+		if parent.Spec.TargetIngress.Name != "" {
+			target, status.JWTAudiences, err = getTargetIngress(parent)
+			if err != nil { //either there is a fatal error or the ingress isn't ready
```

Comment:
getTargetIngress should only return a non-nil error if and only if there is a fatal error that needs to be returned from sync. Please update this block and the getTargetIngress function to match that.